arenadata/db
# Fixing costs for SubqueryScan nodes on queries with CTE when using Postgres optimizer.

PR consist of two parts. One fixes SubqueryScan node cost, another one - plans costs comparison logic.
_______________
Part 1. SubqueryScan node cost patch.

Unlike Postgres, GPDB doesn't use `CteScan` nodes replacing them with `SubqueryScan` nodes. As a part of this feature, much of logic realted to `CteScan` nodes was [changed](https://github.com/greenplum-db/gpdb/commit/41c3b698616bd984b64629c5bae9c0549f07b8ff). One of such changes is [commenting](https://github.com/greenplum-db/gpdb/blob/41c3b698616bd984b64629c5bae9c0549f07b8ff/src/backend/optimizer/plan/planner.c#L732) `SS_process_ctes()` function, which, as described in its comment, can generate unused initplans. The drawback of this solution is wrong, in most cases unusually small, cost of `SubsqueryScan` nodes placed in the position of `CteScan` nodes. It happens because `SS_finalize_plan()` function not [called](https://github.com/greenplum-db/gpdb/blob/41c3b698616bd984b64629c5bae9c0549f07b8ff/src/backend/optimizer/plan/planner.c#L997) - the call to this function, among other things, increases the costs of current plan by costs taken from initplans.
This patch takes care about subplan costs copied to `SubqueryScan` node if subplan available. Additional regression test shows how Generic plan can be preferred over more optimal Custom one, because of bug. More, it shows the case how Generic plan can be preferred over ORCA Custom plan, before plans costs comparison patch(see second part) was applied.

Files changed:
```
createplan.c
gporca.sql and .out files
```

QA
- Why change existing test in `gporca.sql` with `width=0`?
- That's true I can add copying width value from `best_path` below [changes](https://github.com/arenadata/gpdb/pull/234/files#diff-4f0575423ee0e1a3434c5af37bdc873165983c0884489e5ae2c71250c66a0ec3R2958) in `createplan.c`, but I think this just hides the real problem with costs, rows and width in 6x. Result [node](https://github.com/arenadata/gpdb/blob/84a4943c4a2dc27f27860f3870ca6498a3c00486/src/backend/optimizer/plan/createplan.c#L6406) for example. Or `rows=1` problem from the following script.
```
create table subqscan_test1(i integer);
insert into subqscan_test1 select generate_series(0,10000);
analyze subqscan_test1;

create table subqscan_test2(i integer);
insert into subqscan_test2 select generate_series(0,100);
analyze subqscan_test2;

explain
with t1 as
(select i, count(*) c from subqscan_test1 group by i)
select t.i, (select c from t1 where t1.i = t.i) c
from subqscan_test2 t;

 Gather Motion 3:1  (slice2; segments: 3)  (cost=0.00..13845.81 rows=101 width=4)
   ->  Seq Scan on subqscan_test2 t  (cost=0.00..13845.81 rows=34 width=4)
         SubPlan 1  (slice2; segments: 3)
           ->  Subquery Scan on t1  (cost=0.00..137.05 rows=1 width=8)
                 ->  GroupAggregate  (cost=0.00..137.05 rows=1 width=12)
                       Group Key: subqscan_test1.i
                       ->  Result  (cost=0.00..137.02 rows=1 width=4)
                             Filter: (subqscan_test1.i = t.i)
                             ->  Materialize  (cost=0.00..137.02 rows=1 width=4)
                                   ->  Broadcast Motion 3:3  (slice1; segments: 3)  (cost=0.00..137.01 rows=1 width=4)
                                         ->  Seq Scan on subqscan_test1  (cost=0.00..137.01 rows=1 width=4)
```
This should be discussed.

- Why fix [here](https://github.com/arenadata/gpdb/pull/234/files#diff-4f0575423ee0e1a3434c5af37bdc873165983c0884489e5ae2c71250c66a0ec3R2957) and not [here](https://github.com/greenplum-db/gpdb/blob/b6024c99b6a89b33cf12d68cf5afce2585de8555/src/backend/optimizer/util/pathnode.c#L3111) like in master branch? 
- There was a huge job done started in higher versions of Postgres with this [commit](https://github.com/greenplum-db/gpdb/commit/41c3b698616bd984b64629c5bae9c0549f07b8ff). And with [modifications](https://github.com/greenplum-db/gpdb/commit/38d88155520790b268b1a380caa9a2d9508edc66) in GP. There is [another](https://github.com/greenplum-db/gpdb/commit/c5f6dbbe84c43741db1a1c6e31bac1e3bb37bfaa) patch we've mentioned on voice chat. It is not directly related to our problem, but I need to show how much differences between 6x and master. Anyway, looking to all code around paths and nodes in 6x and master, in my opinion, the fix is on the right place.
_______________
Part 2. Plans costs comparison patch.

GPORCA doesn't support Param nodes, so it will always fall back to the Postgres planner. What happened without this patch was that the backend code would compare the cost of the custom plan generated with GPORCA with the cost of a generic plan generated with the Postgres planner, and that doesn't make much sense because the GPORCA has a very different cost model from the Postgres planner.

This commit is a rework of 09aa23d32c433dfbf9675286799027f5867484b9 with additional `is_custom_cost_safe` flag. It indicates is it safe to compare costs of custom and generic plan (true value); was custom plan ever been generated using ORCA (false value).
Unlike mentioned commit, this one contains two additional tests, which proves
 - ORCA's custom plan is always chosen instead of Postgres generic plan
 - Switch to generic plan is possible even if ORCA is enabled (fallback case)

Files changed:
```
CTranslatorScalarToDXL.cpp
plancache.c
plancache.h
plancache.sql and .out files
```

QA
- Why `init_plangen_used` is equal to `PLANGEN_PLANNER` by default?
- Generally, it doesn't matter what value used by default, because `init_plangen_used` value starts to mean something only after first call to `BuildCachedPlan()` - the place where the first plan appears and where all comments contains 'plan' word begins to make sense. Another solution was to make `init_plangen_used` a pointer, but we can't create a pointer just directly to `enum`, we need some value to be initialized first. As it's, imo, not clear and not safe to have a pointer to statement's `planGen`, I decided to choose first solution.
__
- Why we ever need this in current PR?
- SubqueryScan node costs bug show how Postgres Generic plan can be preferred over ORCA Custom plan, which is incorrect behavior.
